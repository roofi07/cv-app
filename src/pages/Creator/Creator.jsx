import React from 'react'
import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'
import PropTypes from 'prop-types'
import Immutable from 'immutable'
import ImmutablePropTypes from 'react-immutable-proptypes'
import './Creator.scss'
import CreatorModule from './CreatorModules/CreatorModule'
import { Modules } from './CreatorModules/modules'

export class Creator extends React.Component {
  static propTypes = {
    getCreator: PropTypes.func,
    modulesData: ImmutablePropTypes.list,
    setCreator: PropTypes.func,
  }

  constructor() {
    super()

    this.removeModule = this.removeModule.bind(this)
    this.updateModule = this.updateModule.bind(this)
    this.save = this.save.bind(this)
    this.reset = this.reset.bind(this)

    this.state = {
      modulesData: Immutable.List(),
    }
  }

  async componentWillMount() {
    const { getCreator } = this.props

    await getCreator()

    const { modulesData } = this.props

    this.setState({ modulesData })
  }

  removeModule(id) {
    let { modulesData } = this.state

    modulesData = modulesData.filter(a => a.get('id') !== id)

    this.setState({ modulesData })
  }

  updateModule(moduleData) {
    let { modulesData } = this.state

    modulesData = modulesData.update(
      modulesData.findIndex(a => a.get('id') === moduleData.get('id')),
      a => a.set('data', moduleData.get('data'))
    )

    this.setState({ modulesData })
  }

  save() {
    const { modulesData } = this.state

    this.props.setCreator(modulesData)
  }

  reset() {
    const { modulesData } = this.props

    this.setState({ modulesData })
  }

  addModule(type) {
    let { modulesData } = this.state

    modulesData = modulesData.push(Immutable.Map({
      data: Immutable.Map(),
      id: new Date().valueOf(),
      type,
    }))

    this.setState({ modulesData })
  }

  render() {
    const { modulesData } = this.state

    return (
      <div className="creator">
        <div className="creator__cv">
          {!modulesData.size && (
            <div className="creator__cv-empty">
              Add your first data row! You can choose one of them on right side.
            </div>
          )}
          {modulesData.map(m => (
            <CreatorModule
              key={m.get('id')}
              moduleData={m}
              onRemove={this.removeModule}
              onUpdate={this.updateModule}
            />
          ))}
        </div>
        <div className="creator__tools">
          <div className="creator__actions">
            <button className="creator__action-btn creator__action-btn--default" onClick={this.save}>Save</button>
            <button className="creator__action-btn creator__action-btn--secondary" onClick={this.reset}>Reset</button>
          </div>
          <h3>Tools</h3>
          {Object.keys(Modules).map(a => (
            <button className="creator__tool-btn" key={a} onClick={() => this.addModule(Modules[a])}>
              <i className="fa fa-plus" /> Add {Modules[a]}
            </button>
          ))}
        </div>
      </div>
    )
  }
}

const mapState = ({ creator }) => ({
  modulesData: creator.modulesData,
})

const mapDispatch = ({ creator: { getCreator, setCreator } }) => ({
  getCreator,
  setCreator,
})

export default withRouter(connect(mapState, mapDispatch)(Creator))
