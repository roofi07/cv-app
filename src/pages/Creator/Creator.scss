.creator {
  padding-right: 300px;

  &__cv {
    width: 1024px;
    min-height: 1000px;
    margin: 40px auto;
    padding: 80px;
    border-left: 3px solid #5aa4ff;
    background: #fff;
  }

  &__cv-empty {
    padding: 40px;
    text-align: center;
    color: #999;
  }

  &__tools {
    position: fixed;
    top: 0;
    right: 0;
    bottom: 0;
    width: 300px;
    padding: 20px;
    background-color: #ddd;
  }

  &__actions {
    text-align: right;
  }

  &__action-btn {
    display: inline-block;
    min-width: 80px;
    margin: 10px 10px 0 0;
    padding: 10px;
    font-size: 12px;
    border: 0;
    outline: 0;
    cursor: pointer;

    &:hover {
      opacity: .8;
    }

    &--default {
      background-color: #5aa4ff;      
    }

    &--secondary {
      background-color: #aaa;      
    }
  }

  &__tool-btn {
    display: block;
    width: 100%;
    margin: 10px 0;
    padding: 14px 10px;
    text-align: left;
    background-color: #aaa;
    border: 0;
    outline: 0;
    cursor: pointer;

    &:hover {
      background-color: #bbb;
    }

    i {
      margin-right: 6px;
    }
  }
}
