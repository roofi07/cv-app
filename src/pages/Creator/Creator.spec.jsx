import React from 'react'
import { shallow } from 'enzyme'
import Immutable from 'immutable'
import { Creator } from './Creator'

const data = [
  {
    data: {
      key: 'Test',
      value: 'test',
    },
    id: 1,
    type: 'TextRow',
  },
]

const initPage = overrides => {
  const mockMethods = {
    getCreator: jest.fn(),
    setCreator: jest.fn(),
  }
  const wrapper = shallow(<Creator
    modulesData={Immutable.fromJS(data)}
    {...mockMethods}
    {...overrides}
  />)
  return { mockMethods, wrapper }
}

describe('Creator', () => {
  it('should render as expected', () => {
    const { wrapper } = initPage()
    expect(wrapper).toMatchSnapshot()
  })
})
