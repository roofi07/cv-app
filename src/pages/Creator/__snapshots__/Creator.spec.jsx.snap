// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Creator should render as expected 1`] = `
<div
  className="creator"
>
  <div
    className="creator__cv"
  >
    <div
      className="creator__cv-empty"
    >
      Add your first data row! You can choose one of them on right side.
    </div>
  </div>
  <div
    className="creator__tools"
  >
    <div
      className="creator__actions"
    >
      <button
        className="creator__action-btn creator__action-btn--default"
        onClick={[Function]}
      >
        Save
      </button>
      <button
        className="creator__action-btn creator__action-btn--secondary"
        onClick={[Function]}
      >
        Reset
      </button>
    </div>
    <h3>
      Tools
    </h3>
    <button
      className="creator__tool-btn"
      key="SUB_HEADER"
      onClick={[Function]}
    >
      <i
        className="fa fa-plus"
      />
       Add 
      SubHeader
    </button>
    <button
      className="creator__tool-btn"
      key="TEXT_ROW"
      onClick={[Function]}
    >
      <i
        className="fa fa-plus"
      />
       Add 
      TextRow
    </button>
  </div>
</div>
`;
